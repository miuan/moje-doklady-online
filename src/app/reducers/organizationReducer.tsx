import { createSlice } from '@reduxjs/toolkit'

export const organizationSlice = createSlice({
  name: 'organization',
  initialState: {
      selected: null
  },
  reducers: {
    select: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.selected = action.payload
    },
  },
})

// Action creators are generated for each case reducer function
export const { select } = organizationSlice.actions

export default organizationSlice.reducer