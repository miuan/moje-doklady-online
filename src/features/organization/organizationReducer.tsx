import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { loader } from 'graphql.macro'
import { RootState } from '../store'
import { activeUserId } from '../user/userSlice'

export const ALL_ORGANIZATION_QUERY = loader('./graphql/allOrganizations.gql')

export const loadAll = createAsyncThunk('organizations/load-all', async (payload, { getState, extra: { apolloClient }, rejectWithValue }: any) => {
  const userId = activeUserId(getState())
  if (!userId) rejectWithValue('no user')

  const { data } = await apolloClient.query({
    query: ALL_ORGANIZATION_QUERY,
    variables: { filter: { AND: [{ user_every: { id: userId } }] } },
  })

  return data.allOrganization
}) as any

export const organizationSlice = createSlice({
  name: 'organizations',
  initialState: {
    initializing: false,
    selected: null,
    all: [],
  },
  reducers: {
    select: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.selected = action.payload
    },
    setAll: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.all = action.payload
    },
  },
  extraReducers: {
    [loadAll.fulfilled]: (state: any, action: any) => {
      state.all = action.payload
    },
  },
})

export const organizations = (state: RootState) => state?.organizations
export const organizationsAll = (state: RootState) => organizations(state)?.all || []
export const organizationsInitializing = (state: RootState) => organizations(state)?.initializing
export const organizationsSelected = (state: RootState) => (organizationsAll(state)[0] as any) || ({} as any)
export const organizationsSelectedId = (state: RootState) => organizationsSelected(state)?.id

// Action creators are generated for each case reducer function
export const { select, setAll } = organizationSlice.actions

export default organizationSlice.reducer
