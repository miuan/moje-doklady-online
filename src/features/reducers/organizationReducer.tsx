import { createSlice } from '@reduxjs/toolkit'

export const organizationSlice = createSlice({
  name: 'organization',
  initialState: {
      selected: null,
      all:[]
  },
  reducers: {
    select: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.selected = action.payload
    },
    setAll: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.all = action.payload
    },
  },
})

// Action creators are generated for each case reducer function
export const { select, setAll } = organizationSlice.actions

export default organizationSlice.reducer