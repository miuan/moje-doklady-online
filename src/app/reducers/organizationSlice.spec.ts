import organizationReducer, {
  OrganizationState,
  add,
  update,
  select
} from './organizationSlice';

describe('organization reducer', () => {
  const initialState: OrganizationState = {
    selected: {id: '1', name:'1'},
    all: [{id: '1', name:'1'}, {id: '2', name:'2'}],
  };
  
  it('should handle initial state', () => {
    expect(organizationReducer(undefined, { type: 'unknown' })).toEqual({
      all: []
    });
  });

  it('should add', () => {
    expect(organizationReducer(initialState, add({id:'3', name:'3'}))).toMatchSnapshot()
  });

  it('should update', () => {
    expect(organizationReducer(initialState, update({id:'1', name:'1-updated'}))).toMatchSnapshot()
  });

  it('should select', () => {
    expect(organizationReducer(initialState, select('2'))).toMatchSnapshot()
  });
});
