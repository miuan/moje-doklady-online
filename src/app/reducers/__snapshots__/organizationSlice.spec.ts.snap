// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organization reducer should add 1`] = `
Object {
  "all": Array [
    Object {
      "id": "1",
      "name": "1",
    },
    Object {
      "id": "2",
      "name": "2",
    },
    Object {
      "id": "3",
      "name": "3",
    },
  ],
  "selected": Object {
    "id": "1",
    "name": "1",
  },
}
`;

exports[`organization reducer should select 1`] = `
Object {
  "all": Array [
    Object {
      "id": "1",
      "name": "1",
    },
    Object {
      "id": "2",
      "name": "2",
    },
  ],
  "selected": Object {
    "id": "2",
    "name": "2",
  },
}
`;

exports[`organization reducer should update 1`] = `
Object {
  "all": Array [
    Object {
      "id": "1",
      "name": "1-updated",
    },
    Object {
      "id": "2",
      "name": "2",
    },
  ],
  "selected": Object {
    "id": "1",
    "name": "1",
  },
}
`;
